File: SpiralTraversal.java
// Time Complexity : O(n*m) : m - rows, n - columns
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
// Maintain top left bottom and right pointers, keeping traversing and updating the pointers, which indirectly helps
// us look at the inner square each time

class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> result = new ArrayList<>();
        int m = matrix.length;
        int n = matrix[0].length;
        int top = 0;
        int right = n - 1;
        int bottom = m - 1;
        int left = 0;

        while (left <= right && top <= bottom){
            //top boundary
            for ( int j = left ; j<= right; j++){
                result.add(matrix[top][j]);
            }
            top++;
            //right boundary
            for ( int i = top; i<=bottom ; i++){
                result.add(matrix[i][right]);
            }
            right--;
            if (top <= bottom)
                //bottom boundary
                for ( int j = right ; j>= left; j--){
                    result.add(matrix[bottom][j]);
                }
                bottom--;
            //left boundary
            if (left <= right){
                for ( int i = bottom; i>=top ; i--){
                    result.add(matrix[i][left]);
                }
                left++;
            }
        }
        return result;
    }
}

File: SpiralTraversal.cpp
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> result;
        if(matrix.size() == 0) return result;
        int top = 0;
        int bottom = matrix.size() - 1;
        int left = 0;
        int right = matrix[0].size() - 1;
        
        while (left <= right && top <= bottom)
        {
            for (int i = left; i <= right; i++)
            {
                result.push_back(matrix[top][i]);
            }
            top++;
            
            if (top <= bottom)
            {
                for (int i = top; i <= bottom; i++)
                {
                    result.push_back(matrix[i][right]);
                }
                right--;
            }
            
            if(left <= right && top <= bottom)
            {
                for (int i = right; i >= left; i--)
                {
                    result.push_back(matrix[bottom][i]);   
                }
                bottom--;
            }
            if (left <= right && top <= bottom)
            {
                for (int i = bottom; i >= top; i--)
                {
                    result.push_back(matrix[i][left]); 
                }
                left++;
            }
        }
        return result;
    }
};
